core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2287,6 +2287,22 @@ private[spark] object Utils extends Logging with SparkClassUtils {
     }.map(threadInfoToThreadStackTrace)
   }
 
+  /** Return a heap dump. Used to capture dumps for the web UI */
+  def getHeapHistogram(): Array[String] = {
+    val pid = String.valueOf(ProcessHandle.current().pid())
+    val builder = new ProcessBuilder("jmap", "-histo:live", pid)
```

Comment:
If we get error from executing it, can we get the error back?

`executeAndGetOutput` seems can handle it?

Reply:
Which error do you have in your mind? At this layer, we don't handle it like `getThreadDump`. For example, `getThreadDump` can throw `SecurityException` and `UnsupportedOperationException`, but we ignore them.

Reply:
The error message will go to `stderr` of Executor log.

Reply:
I took a look at `executeAndGetOutput`, but it's the same. It throws `SparkException` instead of get the error back.

https://github.com/apache/spark/blob/7398e93a2be226d0e71af5b632fa640b37ea5e43/core/src/main/scala/org/apache/spark/util/Utils.scala#L1378-L1381

Reply:
should it respect `JAVA_HOME` then `PATH`?

Reply:
We are inside `JVM` already.
```
$ docker run -it --rm openjdk:11 bash
root@d7d08a90ec41:/# jshell
Jun 23, 2023 6:30:14 AM java.util.prefs.FileSystemPreferences$1 run
INFO: Created user preferences directory.
|  Welcome to JShell -- Version 11.0.16
|  For an introduction type: /help intro

jshell> var p = new ProcessBuilder("jmap", "-histo:live", String.valueOf(ProcessHandle.current().pid())).start()
p ==> Process[pid=77, exitValue="not exited"]

jshell> var r = new BufferedReader(new InputStreamReader(p.getInputStream()))
r ==> java.io.BufferedReader@5fa7e7ff

jshell> var line = ""
line ==> ""

jshell> while (line != null) { line = r.readLine(); System.out.println(line); }
 num     #instances         #bytes  class name (module)
-------------------------------------------------------
   1:          8938         412040  [B (java.base@11.0.16)
   2:          8705         208920  java.lang.String (java.base@11.0.16)
```

Reply:
And, the child process inherits the current process's environment.
- https://docs.oracle.com/javase/8/docs/api/java/lang/ProcessBuilder.html
> an environment, which is a system-dependent mapping from variables to values. The initial value is a copy of the environment of the current process (see [System.getenv()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getenv--)).

Reply:
@dongjoon-hyun I mean where to find `jmap`, not environment variables propagation.

Consider the following case:

There is a chance that the multi JDK installed in the machine, let's say we have 8 and 11 install at
```
/opt/openjdk-8
/opt/openjdk-11
```
and commands of 8 are added into `PATH`, `PATH=/opt/openjdk-8/bin:$PATH`.

If the Spark executor process runs with `JAVA_HOME=/opt/openjdk-11`, with such invocation, even the sub-process know `JAVA_HOME` but still search `jmap` from `PATH`, then `/opt/openjdk-8/bin/jmap` will be used.

```
new ProcessBuilder("jmap", ...)
```

Reply:
Like you pointed, there is no problem to run `jmap`, isn't it, @pan3793 ? It's just CLI command without any format change.

Reply:
I suppose the `$JAVA_HOME/bin/jmap`(in above case, should be `/opt/openjdk-11/bin/jmap` instead of `/opt/openjdk-8/bin/jmap`) should be used, instead of the one first present in `PATH`.

And if only `JAVA_HOME` is set, but no `jmap` in `PATH`, the invocation will fail.

Reply:
@pan3793 . Here, the contact on `jmap` is just one of available binary, not the same JDK's `jmap`.

However, I'm interested in this case. Could you make a small reproducer with Docker image?
> Spark executor proccess respect JAVA_HOME to find $JAVA_HOME/bin/java but the subprocess can not find jmap even it knows where JAVA_HOME is.

Reply:
@dongjoon-hyun This is an issue - we should use `$JAVA_HOME/bin/jmap`, not the first `jmap` which happens to be in the `PATH`. It is common to override `JAVA_HOME` to specify the java version to be used explicitly (or even to not have jdk in the PATH at all).

Reply:
Got it. Let me take a look at that perspective, @mridulm and @pan3793 .

In the worst case, I'll limit this feature to K8s only environment because this was initially developed for K8s environment first.